lagranlucha-client/Assets/_Main/Scripts/Game/UI/NumberCounter.cs

```diff
@@ -0,0 +1,155 @@
+        public float Value
+        {
+            get
+            {
+                if (!values.ContainsKey(id))
+                    values.Add(id, persistent ? PlayerPrefs.GetFloat(id, startingValue) : startingValue);
+
+                return values[id];
+            }
+            set
+            {
+                if (!values.ContainsKey(id))
+                    values.Add(id, persistent ? PlayerPrefs.GetFloat(id, startingValue) : startingValue);
+
+                values[id] = value;
+            }
+        }
+        private Action<float> OnValueChange
+        {
+            get
+            {
+                if (!localEvents.ContainsKey(id))
+                    localEvents.Add(id, null);
+
+                return localEvents[id];
+            }
+            set
+            {
+                if (!localEvents.ContainsKey(id))
+                    localEvents.Add(id, null);
+
+                localEvents[id] = value;
+            }
+        }
+
+        #endregion
+
+        #region BEHAVIORS
+
+        private void Awake()
+        {
+            number = GetComponent<TMP_Text>();
+            Value = startingValue;
+            number.text = ValueString;
+            OnValueChange += ValueChanged;
+        }
+
+        private void OnDestroy()
+        {
+            OnValueChange -= ValueChanged;
+        }
+
+        private void SetUI()
+        {
+            number.text = ValueString;
+            if (scaleTween != null)
+                scaleTween.Kill(true);
+
+            scaleTween = transform.DOPunchScale(Vector3.one * ScaleAmount, ScaleDuration);
+        }
+
+        private void OnValidate()
+        {
+            localEvents = new Dictionary<string, Action<float>>();
+            values = new Dictionary<string, float>();
+            number = GetComponent<TMP_Text>();
+            number.text = ValueString;
+        }
+
+        public void Reset()
+        {
+            values.Remove(id);
+            OnValueChange?.Invoke(Value);
+        }
+
+        private void ValueChanged(float newValue)
+        {
+            Value = newValue;
+            SetUI();
+
+            if (persistent && PlayerPrefs.GetFloat(id) != Value)
+                PlayerPrefs.SetFloat(id, Value);
+        }
+
+        public void Increase(float amount = 1)
+        {
+            if (amount == 0)
+                return;
+
+            OnValueChange?.Invoke(Value + amount);
+        }
+
+        public void Decrease(float amount = 1)
+        {
+            Increase(-amount);
+        }
+
+        public void Set(float newValue)
+        {
+            OnValueChange?.Invoke(newValue);
+        }
+
+        public void Subscribe(Action<float> action)
+        {
+            action?.Invoke(Value);
+            OnValueChange += action;
+        }
+
+        public void Unsubscribe(Action<float> action)
+        {
+            if (this == null)
```

Comment:
If `this` is `null` would't this give a `NullReferenceException` instead of entering this code?

Reply:
When stopping unity some OnDestroy run before others and some elements try to unsubscribe when this object is already being destroyed, this fix it.

Reply:
Oh! nice to know :D